File: examples/experiments/multilang/nodejs/thumbnailer.js
const ONE = 1 << 16;
const DEN = 4294967296;
const HALF_DEN = 2147483648;

function intParam(params, name, defaultValue, minimum, maximum) {
  const raw = params && params[name] !== undefined ? Number(params[name]) : defaultValue;
  const value = Number.isFinite(raw) ? Math.trunc(raw) : defaultValue;
  return Math.max(minimum, Math.min(maximum, value));
}

function generateRgb(width, height) {
  const image = Buffer.allocUnsafe(width * height * 3);
  let pos = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      image[pos] = (x * 13 + y * 7 + (x ^ y)) & 0xff;
      image[pos + 1] = (x * 3 + y * 17 + x * y) & 0xff;
      image[pos + 2] = (x * 19 + y * 5 + (x + y) * 11) & 0xff;
      pos += 3;
    }
  }
  return image;
}

function axisMap(srcSize, dstSize) {
  if (dstSize <= 1) {
    return [[0, 0, 0]];
  }
  const scale = Math.floor(((srcSize - 1) * ONE) / (dstSize - 1));
  const result = new Array(dstSize);
  for (let d = 0; d < dstSize; d++) {
    const fp = d * scale;
    const lo = Math.floor(fp / ONE);
    const frac = fp % ONE;
    const hi = lo + 1 < srcSize ? lo + 1 : lo;
    result[d] = [lo, hi, frac];
  }
  return result;
}

function resizeBilinear(src, srcWidth, srcHeight, dstWidth, dstHeight) {
  const xmap = axisMap(srcWidth, dstWidth);
  const ymap = axisMap(srcHeight, dstHeight);
  const out = Buffer.allocUnsafe(dstWidth * dstHeight * 3);
  let outPos = 0;
  for (const [y0, y1, fy] of ymap) {
    const wy0 = ONE - fy;
    const wy1 = fy;
    const row0 = y0 * srcWidth * 3;
    const row1 = y1 * srcWidth * 3;
    for (const [x0, x1, fx] of xmap) {
      const wx0 = ONE - fx;
      const wx1 = fx;
      const w00 = wx0 * wy0;
      const w10 = wx1 * wy0;
      const w01 = wx0 * wy1;
      const w11 = wx1 * wy1;
      const p00 = row0 + x0 * 3;
      const p10 = row0 + x1 * 3;
      const p01 = row1 + x0 * 3;
      const p11 = row1 + x1 * 3;
      for (let channel = 0; channel < 3; channel++) {
        const weighted =
          src[p00 + channel] * w00 + src[p10 + channel] * w10 + src[p01 + channel] * w01 + src[p11 + channel] * w11;
        out[outPos] = Math.floor((weighted + HALF_DEN) / DEN);
        outPos += 1;
      }
    }
  }
  return out;
}

function fnv1a32(data) {
  let value = 2166136261 >>> 0;
  for (const byte of data) {
    value ^= byte;
    value = Math.imul(value, 16777619) >>> 0;
  }
  return value >>> 0;
}

module.exports = async function handler(params, context) {
  params = params || {};
  const srcWidth = intParam(params, "src_width", 1024, 64, 4096);
  const srcHeight = intParam(params, "src_height", 768, 64, 4096);
  const dstWidth = intParam(params, "dst_width", 320, 16, srcWidth);
  const dstHeight = intParam(params, "dst_height", 240, 16, srcHeight);
  const rounds = intParam(params, "rounds", 8, 1, 50);
  const source = generateRgb(srcWidth, srcHeight);
  let resized = Buffer.alloc(0);
  for (let round = 0; round < rounds; round++) {
    resized = resizeBilinear(source, srcWidth, srcHeight, dstWidth, dstHeight);
  }
  return {
    benchmark: "thumbnailer",
    source_width: srcWidth,
    source_height: srcHeight,
    thumbnail_width: dstWidth,
    thumbnail_height: dstHeight,
    rounds,
    processed_pixels: dstWidth * dstHeight * rounds,
    output_bytes: resized.length,
    checksum: fnv1a32(resized),
  };
};

File: examples/experiments/multilang/python/thumbnailer.py
MASK32 = (1 << 32) - 1
ONE = 1 << 16
DEN = 1 << 32
HALF_DEN = 1 << 31


def _int_param(params, name, default, minimum, maximum):
    try:
        value = int(params.get(name, default))
    except (TypeError, ValueError, AttributeError):
        value = default
    return max(minimum, min(maximum, value))


def _generate_rgb(width, height):
    image = bytearray(width * height * 3)
    pos = 0

    for y in range(height):
        for x in range(width):
            image[pos] = (x * 13 + y * 7 + (x ^ y)) & 0xFF
            image[pos + 1] = (x * 3 + y * 17 + (x * y)) & 0xFF
            image[pos + 2] = (x * 19 + y * 5 + ((x + y) * 11)) & 0xFF
            pos += 3

    return image


def _axis_map(src_size, dst_size):
    if dst_size <= 1:
        return [(0, 0, 0)]

    scale = ((src_size - 1) * ONE) // (dst_size - 1)
    result = []

    for d in range(dst_size):
        fp = d * scale
        lo = fp >> 16
        frac = fp & 0xFFFF
        hi = lo + 1 if lo + 1 < src_size else lo
        result.append((lo, hi, frac))

    return result


def _resize_bilinear(src, src_width, src_height, dst_width, dst_height):
    xmap = _axis_map(src_width, dst_width)
    ymap = _axis_map(src_height, dst_height)
    out = bytearray(dst_width * dst_height * 3)
    out_pos = 0

    for y0, y1, fy in ymap:
        wy0 = ONE - fy
        wy1 = fy
        row0 = y0 * src_width * 3
        row1 = y1 * src_width * 3

        for x0, x1, fx in xmap:
            wx0 = ONE - fx
            wx1 = fx
            w00 = wx0 * wy0
            w10 = wx1 * wy0
            w01 = wx0 * wy1
            w11 = wx1 * wy1

            p00 = row0 + x0 * 3
            p10 = row0 + x1 * 3
            p01 = row1 + x0 * 3
            p11 = row1 + x1 * 3

            for channel in range(3):
                weighted = (
                    src[p00 + channel] * w00
                    + src[p10 + channel] * w10
                    + src[p01 + channel] * w01
                    + src[p11 + channel] * w11
                )
                out[out_pos] = (weighted + HALF_DEN) // DEN
                out_pos += 1

    return out


def _fnv1a32(data):
    value = 2166136261

    for byte in data:
        value ^= byte
        value = (value * 16777619) & MASK32

    return value


def handler(params, context):
    params = params or {}

    src_width = _int_param(params, "src_width", 1024, 64, 4096)
    src_height = _int_param(params, "src_height", 768, 64, 4096)
    dst_width = _int_param(params, "dst_width", 320, 16, src_width)
    dst_height = _int_param(params, "dst_height", 240, 16, src_height)
    rounds = _int_param(params, "rounds", 8, 1, 50)

    source = _generate_rgb(src_width, src_height)
    resized = b""

    for _ in range(rounds):
        resized = _resize_bilinear(source, src_width, src_height, dst_width, dst_height)

    return {
        "benchmark": "thumbnailer",
        "source_width": src_width,
        "source_height": src_height,
        "thumbnail_width": dst_width,
        "thumbnail_height": dst_height,
        "rounds": rounds,
        "processed_pixels": dst_width * dst_height * rounds,
        "output_bytes": len(resized),
        "checksum": _fnv1a32(resized),
    }
